src/library/scala/collection/immutable/List.scala

```diff
@@ -556,6 +556,44 @@ sealed abstract class List[+A]
     result
   }
 
+  override def partition(p: A => Boolean): (List[A], List[A]) = {
+    if (isEmpty) List.TupleOfNil
+    else {
+      val b0 = p(this.head)
+      var canConserve = true
+      var ys = this
+      var ayes: ListBuffer[A] = null
+      var nays: ListBuffer[A] = null
+      var n = 0
+      while (!ys.isEmpty) {
+        val y = ys.head
+        val b = if (n == 0) b0 else p(y)
+        val newCanConserve = canConserve && b == b0
+        if (canConserve && !newCanConserve) {
+          ayes = new ListBuffer[A]
+          nays = new ListBuffer[A]
+          val prefix = if (b0) ayes else nays
+          var j = 0
+          var zs = this
+          while (j < n) {
+            prefix += zs.head
+            zs = zs.tail
+            j += 1
+          }
```

Comment:
In the extreme case (if only the last element of the list needs to go to the other bucket), the entire list is traversed twice. Is that a good trade off?

Reply:
@lrytz Currently, the class `List` inherits its implementation of `partition` from the `Iterable` trait (at least I could not find any `override` in between). The Scaladoc comment for `Iterable.partition` says: 

>   The default implementation provided here needs to traverse the collection twice.
>   Strict collections have an overridden version of `partition` in `StrictOptimizedIterableOps`,
>  which requires only a single traversal.

So this is not a loss.

Reply:
List does inherit `StringOptimizedIterableOps.partition`:

```
Welcome to Scala 2.13.1 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_172).
Type in expressions for evaluation. Or try :help.

scala> :power
Power mode enabled. :phase is at typer.
import scala.tools.nsc._, intp.global._, definitions._
Try :help or completions for vals._ and power._

scala> typeOf[List[_]].member(TermName("partition")).fullLocationString
res0: String = method partition in trait StrictOptimizedIterableOps
```

Reply:
I haven't been able to measure the performance slowdown. `ListBenchmark.partition_exc_last` tests this case. I tried with `-psize=1000000`, as for smaller data sizes the second traversal is often cheaper due to CPU cache effects.

Reply:
Thanks. Still slightly surprised - do you have an intuition?

Reply:
The difference for that worst case is between one traversal and allocating `N` new `::` objects, or two traversals ans allocating `N-1` objects. It may be that the cost of the extra traversal is negligible next to that of the allocations.

To put it in another way, for each element we have: 
- Before: load  `::` object, load element, apply function to element, create new `::` object.
- After: load `::` object, load element, apply function to element; and for the second iteration (clone prefix), re-load `::` object, and create new `::` object. 

It is only the second re-loading of the `::` object, per iteration, (from the source list) that changes from one method to the other. My intuition would be, this is negligible next to the rest.